### PR TITLE
SONAR-5856 Rewrite Settings page

### PR DESCRIPTION
**Pull request by stas-vilchik**

### CODE REVIEW

**Comment by julienlancelot** [comment]
Here are my feedbacks : 
- When clicking on some category, sometimes nothing is displayed and the following error is displayed in the JS console (I think it's coming from a plugin I'm using from IT : property-sets-plugin) : 
```
TypeError: definition is undefined
return 'settings_' + definition.key + '_' + (0, _uniqueId2.default)('setting_')
```
- There seems to have a kind of caches issues : some settings from other categories are always displayed

**Comment by julienlancelot** [comment]
The "Key: ..." line should be in another color than the description

**Comment by julienlancelot** [comment]
Otherwise it looks very good !!!

File: server/sonar-server/src/main/java/org/sonar/server/settings/ws/ValuesAction.java [hunk 1]
```diff
@@ -0,0 +1,213 @@
...
+
+    private void processDefinitions() {
+      propertyDefinitions.getAll().stream()
+        .filter(definition -> keys.contains(definition.key()))
+        .filter(defaultProperty -> !isNullOrEmpty(defaultProperty.defaultValue()))
+        .map(Setting::new)
+        .forEach(this::processDefaultValue);
+    }
+
+    private void processDefaultValue(Setting setting) {
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/settings/ws/ValuesAction.java]: MAJOR Remove this unused private "processDefaultValue" method. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AUnusedPrivateMethod)

File: server/sonar-web/src/main/js/apps/settings/components/inputs/renderInput.js [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+    TYPE_BOOLEAN,
+    TYPE_INTEGER,
+    TYPE_LONG,
+    TYPE_TEXT,
+    TYPE_PASSWORD,
+    TYPE_FLOAT,
+    TYPE_SINGLE_SELECT_LIST
+} from '../../constants';
+
+const renderInput = setting => {
```
Comment (SonarTech) [inline server/sonar-web/src/main/js/apps/settings/components/inputs/renderInput.js]: MAJOR Function has a complexity of 11 which is greater than 10 authorized. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=javascript%3AFunctionComplexity)

File: sonar-db/src/main/java/org/sonar/db/property/PropertiesDao.java [hunk 3]
```diff
@@ -161,6 +164,23 @@ public PropertyDto selectProjectProperty(long resourceId, String propertyKey) {
...
+  public List<PropertyDto> selectPropertiesByKeysAndComponentId(DbSession session, Set<String> keys, long componentId) {
+    return selectByKeys(session, keys, componentId);
+  }
+
+  public List<PropertyDto> selectPropertiesByKeysAndComponentIds(DbSession session, Set<String> keys, Set<Long> componentIds) {
+    return executeLargeInputs(keys, partitionKeys -> executeLargeInputs(componentIds,
+      partitionComponentIds -> session.getMapper(PropertiesMapper.class).selectByKeysAndComponentIds(partitionKeys, partitionComponentIds)));
+  }
+
+  private List<PropertyDto> selectByKeys(DbSession session, Set<String> keys, @Nullable Long componentId) {
```
Comment (SonarTech) [inline sonar-db/src/main/java/org/sonar/db/property/PropertiesDao.java]: MAJOR Make "selectByKeys" a "static" method. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS2325)

**Comment by SonarTech** [comment]
SonarQube analysis reported 3 issues
* MAJOR 3 major

Watch the comments in this conversation to review them.

File: server/sonar-web/src/main/js/apps/settings/components/inputs/renderInput.js [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+    TYPE_BOOLEAN,
+    TYPE_INTEGER,
+    TYPE_LONG,
+    TYPE_TEXT,
+    TYPE_PASSWORD,
+    TYPE_FLOAT,
+    TYPE_SINGLE_SELECT_LIST
+} from '../../constants';
+
+const renderInput = (setting, onSet) => {
```
Comment (SonarTech) [inline server/sonar-web/src/main/js/apps/settings/components/inputs/renderInput.js]: MAJOR Function has a complexity of 11 which is greater than 10 authorized. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=javascript%3AFunctionComplexity)

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue
* MAJOR 1 major

Watch the comments in this conversation to review them.

File: server/sonar-web/src/main/js/apps/settings/components/DefinitionDefaults.js [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+import React from 'react';
+import { getSettingValue, isEmptyValue, isComplexDefinition, isDefaultOrInherited } from '../utils';
+import { translate } from '../../../helpers/l10n';
+
+// TODO display value to which the setting will be reset
```
Comment (SonarTech) [inline server/sonar-web/src/main/js/apps/settings/components/DefinitionDefaults.js]: INFO Complete the task associated to this TODO comment. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=javascript%3AS1135)

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue
* INFO 1 info

Watch the comments in this conversation to review them.

File: server/sonar-web/src/main/js/apps/settings/utils.js [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+    return setting.parentValue;
+  }
+}
+
+/**
+ * Get and format the default value
+ * @param setting
+ * @returns {string}
+ */
+export function getDefaultValue (setting) {
```
Comment (SonarTech) [inline server/sonar-web/src/main/js/apps/settings/utils.js]: MAJOR Function has a complexity of 12 which is greater than 10 authorized. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=javascript%3AFunctionComplexity)

File: server/sonar-web/src/main/js/apps/settings/store/actions.js [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+};
+
+export const saveValue = (key, componentKey) => (dispatch, getState) => {
+  dispatch(startLoading(key));
+
+  const state = getState();
+  const definition = getDefinition(state, key);
+  const value = getChangedValue(state, key);
+
+  // TODO avoid call to #getValues()
```
Comment (SonarTech) [inline server/sonar-web/src/main/js/apps/settings/store/actions.js]: INFO Complete the task associated to this TODO comment. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=javascript%3AS1135)

File: server/sonar-web/src/main/js/apps/settings/store/actions.js [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+        dispatch(passValidation(key));
+        dispatch(stopLoading(key));
+      })
+      .catch(e => parseError(e).then(message => dispatch(failValidation(key, message))));
+};
+
+export const resetValue = (key, componentKey) => dispatch => {
+  dispatch(startLoading(key));
+
+  // TODO avoid call to #getValues()
```
Comment (SonarTech) [inline server/sonar-web/src/main/js/apps/settings/store/actions.js]: INFO Complete the task associated to this TODO comment. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=javascript%3AS1135)

**Comment by SonarTech** [comment]
SonarQube analysis reported 3 issues
* MAJOR 1 major
* INFO 2 info

Watch the comments in this conversation to review them.

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 issues
* MAJOR 2 major

Watch the comments in this conversation to review them.

#### 1 extra issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. MAJOR [license](https://github.com/SonarSource/sonarqube/blob/1b93697fc2a548102ee38e5cb728cf7f5620ec17/sonar-ws/src/main/java/org/sonarqube/ws/client/license): Add a 'package-info.java' file to document the 'license' package [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1228)

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue
* MAJOR 1 major

Watch the comments in this conversation to review them.

**Pull request merged**